### ISSUE TITLE
Exception in thread "main" scala.MatchError: [Ljava.lang.String;@4a94ee4 (of class [Ljava.lang.String;)

### ISSUE THREAD

**Issue by germal** [PacktPublishing/Hands-On-Deep-Learning-with-Apache-Spark#1]
Hi Guglielmo
I am trying the code in the chapter 4.
I am not a Java or Scala developer and I'm feeling loss with the errors that I found executing the jar that I have compiled with Maven project.
I appended to the pom.xml file the scala plugin needed for scala code build and a shaded plugin needed to solve an error about Hadoop filesystem not found .The following is the build part that I have appended

# <build>
#        <plugins>
#            
#            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.8.0</version>
                <configuration>
                    <source>${java.version}</source>
                    <target>${java.version}</target>
                </configuration>
            </plugin>

            
            <plugin>
                <groupId>org.codehaus.mojo</groupId>
                <artifactId>build-helper-maven-plugin</artifactId>
                <version>3.0.0</version>
                <executions>
                    <execution>
                        <id>add-source</id>
                        <phase>generate-sources</phase>
                        <goals>
                            <goal>add-source</goal>
                        </goals>
                        <configuration>
                            <sources>
                                <source>src/main/scala</source>
                            </sources>
                        </configuration>
                    </execution>
                    <execution>
                        <id>add-test-source</id>
                        <phase>generate-test-sources</phase>
                        <goals>
                            <goal>add-test-source</goal>
                        </goals>
                        <configuration>
                            <sources>
                                <source>src/test/scala</source>
                            </sources>
                        </configuration>
                    </execution>
                </executions>
            </plugin>

            
            <plugin>
                <groupId>net.alchim31.maven</groupId>
                <artifactId>scala-maven-plugin</artifactId>
                <version>3.2.2</version>
                <executions>
                    <execution>
                        <goals>
                            
                            <goal>compile</goal>
                            <goal>testCompile</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
<plugin>
  <groupId>org.apache.maven.plugins</groupId>
  <artifactId>maven-shade-plugin</artifactId>
  <version>2.3</version>
  <executions>
    <execution>
      <phase>package</phase>
      <goals>
        <goal>shade</goal>
      </goals>
      <configuration>
        <transformers>
          <transformer implementation="org.apache.maven.plugins.shade.resource.ServicesResourceTransformer"/>
        </transformers>
      </configuration>
    </execution>
  </executions>
</plugin>
        </plugins>
</build>  -->

The following is the command line that I use to submit

$SPARK_HOME/bin/spark-submit --master local --class org.googlielmo.sparkstreamingkafka.DirectKafkaDataVec --jars /data/handsonspark/Chapter04/sparkstreamingkafka/target/sparkstreamingkafka-1.0.jar /usr/apache/spark-2.2.3/jars/*

Now I am stuck on this error :
Exception in thread "main" scala.MatchError: [Ljava.lang.String;@4a94ee4 (of class [Ljava.lang.String;)
at org.googlielmo.sparkstreamingkafka.DirectKafkaDataVec$.main(DirectKafkaDataVec.scala:30)
at org.googlielmo.sparkstreamingkafka.DirectKafkaDataVec.main(DirectKafkaDataVec.scala)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:498)
at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:775)
at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:180)
at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:205)
at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:119)
at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)

Please help because I am stuck from 2 days and I need to go on
Regards
germal

**Comment by virtualramblas** [comment]
Hi Germal, 
Looking at the spark-submit command you issued, you have passed the application JAR (_sparkstreamingkafka-1.0.jar_) as part of the _--jars_ option, while the other dependencies are passed as the application jar. Please have a look at the official Spark documentation [page ](https://spark.apache.org/docs/latest/submitting-applications.html) or go back to chapter 1 of the book for the proper syntax of the spark-submit command.

Best Regards,
Guglielmo

**Comment by germal** [comment]
Hi Guglielmo
In your book at chapter 1 it is written 
**$SPARK_HOME/bin/spark-submit --class path.to.your.Class --master yarn --deploy-mode cluster [options] <app jar> [app options]
In cluster mode, since the driver runs on a different machine than the client, the SparkContext.addJar method doesn't work with the files that are local to the client. The only choice is to include them using the jars option in the launch command.**
Executing with the same pattern of the command I get an error of class not found and obmitting the --jars option it doesn't work at all as in the attached screenshot.
Furthermore At chapter 4 it is written 
**Let's build the Spark application and execute it through the $SPARK_HOME/bin/spark-submit command, specifying the JAR filename, the Spark master URL, the job name, the main class name, the maximum memory to be used by each executor, and the job arguments (localhost:9092 and csvtopic).**

Could you please clarify what you mean with " the job arguments (localhost:9092 and csvtopic)." ?
How it is possible to reference the csvtopic in the spark-submint command ?
Thanks 
Germal
![error](https://user-images.githubusercontent.com/47027729/52363201-7b848000-2a42-11e9-9406-97f98e943ba9.jpg)

**Comment by virtualramblas** [comment]
Hi Germal,

Thanks for sharing the screenshot for the two commands you have executed: it really helped me on providing you the proper answer to your questions.
Looking at the console output you shared, I can see that the submission of the Spark job attempted through the second command has been successful (as expected, being that a correct command form): only things missing, as printed to console by the job itself, are its expected arguments (which include the list of Kafka brokers and the topics to consume from). They can be passed to the application as part of the spark-submit command as follows:

_$SPARK_HOME/bin/spark-submit --class org.googlielmo.sparkstreamingkafka.DirectKafkaDataVec --deploy-mode client /data/handsonspark/Chapter04/sparkstreamingkafka/target/sparkstreamingkafka-1.0.jar spark://localhost:7077 localhost:9092 csvtopic_

(please replace the Kafka broker name and port and the topic name with those for your Kafka cluster).

In order to pass the Spark job dependencies, you can use the _--packages_ option of the spark-submit command (please refer to the documentation link (section "Advanced Dependency Management") I shared in my previous reply). This option automatically resolves all of the transitive dependencies too. Alternatively you can build through Maven a shaded fat JAR, as explained in chapter 10.

Best Regards,
Guglielmo

**Comment by germal** [comment]
Hi Guglielmo
Thanks for the help

**root@osboxes:/data/handsonspark/Chapter04/sparkstreamingkafka# $SPARK_HOME/bin/spark-submit --class org.googlielmo.sparkstreamingkafka.DirectKafkaDataVec --deploy-mode client  /data/handsonspark/Chapter04/sparkstreamingkafka/target/sparkstreamingkafka-1.0.jar spark://osboxes:7077 osboxes:9092 packttopic**
But Now I get the following error 
**Exception in thread "main" org.apache.spark.SparkException: java.nio.channels.ClosedChannelException**
Even if I created the topic and started the producer with the command
$KAFKA_HOME/bin/kafka-console-producer.sh --broker-list localhost:9092 --topic packttopic

**The following it's the complete output
![error](https://user-images.githubusercontent.com/47027729/52373979-61579b80-2a5c-11e9-91a8-91e81a646642.jpg)
**
Schema():
idx   name               type           meta data
0     "id"               String         StringMetaData(name="id",)
1     "description"      String         StringMetaData(name="description",)
2     "notes"            String         StringMetaData(name="notes",)




Schema after transforming data:
Schema():
idx   name               type           meta data
0     "id"               String         StringMetaData(name="id",)
1     "description"      String         StringMetaData(name="description",)

Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
19/02/06 15:57:47 INFO SparkContext: Running Spark version 2.2.3
19/02/06 15:57:52 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
19/02/06 15:57:52 WARN Utils: Your hostname, osboxes resolves to a loopback address: 127.0.1.1; using 192.xxxxxxxxxxinstead (on interface ens33)
19/02/06 15:57:52 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
19/02/06 15:57:52 INFO SparkContext: Submitted application: DirectKafkaDataVec
19/02/06 15:57:52 INFO SecurityManager: Changing view acls to: root
19/02/06 15:57:52 INFO SecurityManager: Changing modify acls to: root
19/02/06 15:57:52 INFO SecurityManager: Changing view acls groups to: 
19/02/06 15:57:52 INFO SecurityManager: Changing modify acls groups to: 
19/02/06 15:57:52 INFO SecurityManager: SecurityManager: authentication disabled; ui acls disabled; users  with view permissions: Set(root); groups with view permissions: Set(); users  with modify permissions: Set(root); groups with modify permissions: Set()
19/02/06 15:57:53 INFO Utils: Successfully started service 'sparkDriver' on port 35921.
19/02/06 15:57:53 INFO SparkEnv: Registering MapOutputTracker
19/02/06 15:57:54 INFO SparkEnv: Registering BlockManagerMaster
19/02/06 15:57:54 INFO BlockManagerMasterEndpoint: Using org.apache.spark.storage.DefaultTopologyMapper for getting topology information
19/02/06 15:57:54 INFO BlockManagerMasterEndpoint: BlockManagerMasterEndpoint up
19/02/06 15:57:54 INFO DiskBlockManager: Created local directory at /tmp/blockmgr-64f0c99f-7787-417b-846d-d52cbdef621d
19/02/06 15:57:54 INFO MemoryStore: MemoryStore started with capacity 366.3 MB
19/02/06 15:57:54 INFO SparkEnv: Registering OutputCommitCoordinator
19/02/06 15:57:54 INFO Utils: Successfully started service 'SparkUI' on port 4040.
19/02/06 15:57:54 INFO SparkUI: Bound SparkUI to 0.0.0.0, and started at http://192.1xxxxxxxxx:4040
19/02/06 15:57:54 INFO SparkContext: Added JAR file:/data/handsonspark/Chapter04/sparkstreamingkafka/target/sparkstreamingkafka-1.0.jar at spark://192.1xxxxxxx:35921/jars/sparkstreamingkafka-1.0.jar with timestamp 1549486674971
19/02/06 15:57:55 INFO StandaloneAppClient$ClientEndpoint: Connecting to master spark://osboxes:7077...
19/02/06 15:57:55 INFO TransportClientFactory: Successfully created connection to osboxes/127.0.1.1:7077 after 85 ms (0 ms spent in bootstraps)
19/02/06 15:57:55 INFO StandaloneSchedulerBackend: Connected to Spark cluster with app ID app-20190206155755-0004
19/02/06 15:57:55 INFO StandaloneAppClient$ClientEndpoint: Executor added: app-20190206155755-0004/0 on worker-20190206154649-192.1xxxxxxx-42581 (192.xxxxxxxxxx:42581) with 4 cores
19/02/06 15:57:55 INFO Utils: Successfully started service 'org.apache.spark.network.netty.NettyBlockTransferService' on port 45219.
19/02/06 15:57:55 INFO StandaloneSchedulerBackend: Granted executor ID app-20190206155755-0004/0 on hostPort 192.xxxxxxxxxx:42581 with 4 cores, 1024.0 MB RAM
19/02/06 15:57:55 INFO NettyBlockTransferService: Server created on 192.1xxxxxx:45219
19/02/06 15:57:55 INFO BlockManager: Using org.apache.spark.storage.RandomBlockReplicationPolicy for block replication policy
19/02/06 15:57:55 INFO StandaloneAppClient$ClientEndpoint: Executor updated: app-20190206155755-0004/0 is now RUNNING
19/02/06 15:57:56 INFO BlockManagerMaster: Registering BlockManager BlockManagerId(driver, 192.168.227.152, 45219, None)
19/02/06 15:57:56 INFO BlockManagerMasterEndpoint: Registering block manager 192.168.227.152:45219 with 366.3 MB RAM, BlockManagerId(driver, 192.xxxxx, 45219, None)
19/02/06 15:57:56 INFO BlockManagerMaster: Registered BlockManager BlockManagerId(driver, 192xxxxxxx, 45219, None)
19/02/06 15:57:56 INFO BlockManager: Initialized BlockManager: BlockManagerId(driver, 192.1xxxxxxxxx, 45219, None)
19/02/06 15:57:56 INFO StandaloneSchedulerBackend: SchedulerBackend is ready for scheduling beginning after reached minRegisteredResourcesRatio: 0.0
Exception in thread "main" org.apache.spark.SparkException: java.nio.channels.ClosedChannelException
	at org.apache.spark.streaming.kafka.KafkaCluster$$anonfun$checkErrors$1.apply(KafkaCluster.scala:366)
	at org.apache.spark.streaming.kafka.KafkaCluster$$anonfun$checkErrors$1.apply(KafkaCluster.scala:366)
	at scala.util.Either.fold(Either.scala:98)
	at org.apache.spark.streaming.kafka.KafkaCluster$.checkErrors(KafkaCluster.scala:365)
	at org.apache.spark.streaming.kafka.KafkaUtils$.getFromOffsets(KafkaUtils.scala:222)
	at org.apache.spark.streaming.kafka.KafkaUtils$.createDirectStream(KafkaUtils.scala:484)
	at org.googlielmo.sparkstreamingkafka.DirectKafkaDataVec$.main(DirectKafkaDataVec.scala:57)
	at org.googlielmo.sparkstreamingkafka.DirectKafkaDataVec.main(DirectKafkaDataVec.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:775)
	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:180)
	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:205)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:119)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)

Thanks in advance for the support !
Regards
Germal

**Comment by virtualramblas** [comment]
Hi Germal, 

The first thing comes to my mind is that probably the hostname _osboxes_  for some reason can't be resolved. I have noticed you started the producer on localhost:9092. Did you try using the _osboxes_ machine IP address or localhost when submitting the Spark job?

Best Regards,
Guglielmo

**Comment by germal** [comment]
Hi Guglielmo

Yes that was my mistake in the hosts file !
Please see the following new error that I face 
![error](https://user-images.githubusercontent.com/47027729/52385003-2cf5d680-2a80-11e9-81ba-648c46d4299d.jpg)
Thanks
Germal

**Comment by germal** [comment]
I face the error as soon as I send the first message from the producer
Thanks a lot for your great help !
Regards
Germal

**Comment by virtualramblas** [comment]
Hi Germal,

That's the exception that the TransformProcess class throws when the input data schema's number of fields doesn't match what has been defined programmatically. I believe in your case, because you're inputting data manually, the interval for the Spark StreamingContext is too short (it is set to 5 seconds in the code), so a single line is truncated and consumed before you can complete adding all of the 3 fields for a single record. Try to set a longer interval (line 51 of the example).

Best Regards,
Guglielmo

**Comment by germal** [comment]
Hi Guglielmo
Thank you very much for your explanations , they are crystal clear and help me a lot to understand.
The exception in the last screenshot was due to my misconfiguration using DirectKafkaDataVec with packttopic and not with csvtopic , so the format of the messages mismatched.
Anyway I found a typo in the example dataset i(5 columns) in the book that mismatches with the set format ( 3 columns). So I used a csv with 3 field ,
Thanks a lot for the super support !
![error](https://user-images.githubusercontent.com/47027729/52423708-f314e680-2af8-11e9-846e-ca65db33c5fc.jpg)
